// yahoo-finance2/dist/cjs/src/modules/historical.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const queryOptionsDefaults = {
    interval: "1d",
    events: "history",
    includeAdjustedClose: true,
};
function historical(symbol, queryOptionsOverrides, moduleOptions) {
    return this._moduleExec({
        moduleName: "historical",
        query: {
            url: "https://query1.finance.yahoo.com/v7/finance/download/" + symbol,
            schemaKey: "#/definitions/HistoricalOptions",
            defaults: queryOptionsDefaults,
            overrides: queryOptionsOverrides,
            fetchType: "csv",
            transformWith(queryOptions) {
                if (!queryOptions.period2)
                    queryOptions.period2 = new Date();
                const dates = ["period1", "period2"];
                for (const fieldName of dates) {
                    const value = queryOptions[fieldName];
                    if (value instanceof Date)
                        queryOptions[fieldName] = Math.floor(value.getTime() / 1000);
                    else
                        typeof value === "string";
                    queryOptions[fieldName] = Math.floor(new Date(value).getTime() / 1000);
                }
                return queryOptions;
            },
        },
        result: {
            schemaKey: "#/definitions/HistoricalResult",
            transformWith(result) {
                const filteredResults = [];
                const fieldCount = Object.keys(result[0]).length;
                // Count number of null values in object (1-level deep)
                function nullFieldCount(object) {
                    let nullCount = 0;
                    for (const val of Object.values(object))
                        if (val === null)
                            nullCount++;
                    return nullCount;
                }
                for (const row of result) {
                    const nullCount = nullFieldCount(row);
                    if (nullCount === 0) {
                        // No nulls is a legit (regular) result
                        filteredResults.push(row);
                    }
                    else if (nullCount !== fieldCount - 1 /* skip "date" */) {
                        // Unhandled case: some but not all values are null.
                        // Note: no need to check for null "date", validation does it for us
                        console.error(nullCount, row);
                        throw new Error("Historical returned a result with SOME (but not " +
                            "all) null values.  Please report this, and provide the " +
                            "query that caused it.");
                    }
                    else {
                        // All fields (except "date") are null: silently skip (no-op)
                    }
                }
                /*
                 * We may consider, for future optimization, to count rows and create
                 * new array in advance, and skip consecutive blocks of null results.
                 * Of doubtful utility.
                 */
                return filteredResults;
            },
        },
        moduleOptions,
    });
}
exports.default = historical;
